doc/macro_utils_design.md

```diff
@@ -0,0 +1,240 @@
+
+Assuming "n" is 4, here's some examples of what the output of the macro is
+
+MU_THE_NTH_ARG(1,2,3,4) => 4
+
+MU_THE_NTH_ARG(11, 22, 33, 44, 55) => 44
+
+MU_THE_NTH_ARG(11, 22, 33) => error, the macro requires 4 or more arguments. This is ok, because in all the context where this macro is called, there will be more than 4 arguments passed to it.
+
+`MU_THE_NTH_ARG` is not extremely useful by itself. When used with `MU_COUNT_1_OR_MORE_ARG` it becomes very useful.
+
+### MU_COUNT_1_OR_MORE_ARG(...)
+
+`MU_COUNT_1_OR_MORE_ARG` counts ("expands to a number") the number of the arguments passed to it. In all instances where this macro is called, there must be at least 1 argument passed to it.
+
+Here's some examples:
+
+MU_COUNT_1_OR_MORE_ARG(X) => 1
+
+MU_COUNT_1_OR_MORE_ARG(a,b) => 2
+
+MU_COUNT_1_OR_MORE_ARG(a, ,c) => 3
+
+Here is how that is achieved. We're going to call `MU_THE_NTH_ARG`(<<here are `MU_COUNT_1_OR_MORE_ARG`'s arguments>>, 3,2,1).
+
+Here's how it works for the examples above:
+MU_COUNT_1_OR_MORE_ARG(X) => MU_THE_NTH_ARG(X, 3, 2, 1) => 1.
+                                                     ^
+                                                     |----- this is 4th argument
+
+MU_COUNT_1_OR_MORE_ARG(a,b) => MU_THE_NTH_ARG(a, b, 3, 2, 1) => 2.
+                                                       ^
+                                                       |----- this is 4th argument
+
+MU_COUNT_1_OR_MORE_ARG(a, ,c) => MU_THE_NTH_ARG(a, , c, 3, 2, 1) => 3
+                                                        ^
+                                                        |----- this is 4th argument
+
+So what happens if the macro is called with more than 4 arguments? Such as a,b,c,d,e?
+
+MU_COUNT_1_OR_MORE_ARG(a, ,c) => MU_THE_NTH_ARG(a, b, c, d, e, 3, 2, 1) => d
+                                                         ^
+                                                         |----- this is 4th argument
+
+Obviously `d` is not the expected macro expansion at this moment. So the macro has some limitations, it can count only 1...3 arguments. (it can count 1..n-1, where n is coming from `MU_THE_NTH_ARG`)
+
+So what happens if the macro is called with 0 arguments?
+MU_COUNT_1_OR_MORE_ARG() => MU_THE_NTH_ARG( , 3, 2, 1) => 1
+                                                    ^
+                                                    |----- this is 4th argument
+                                                
+Here we see how the macro does not produce 0, and that is precisely why the macro is called MU_COUNT_**1**_OR_MORE_ARG. 
+
+Next question is : how to produce 0 as a count result? Not easy or immediately stright forward. More macros are needed for that.
+
+### MU_HAS_COMMA(...)
+
+`MU_HAS_COMMA` expands to 0 when the number of its arguments is 0 or 1. If the number of arguments is greater or equal to 2 then it expands to 0. Here' some examples
```

Comment:
>Here' [](start = 146, length = 5)

Here are some

Reply:
grammarified.

---
In reply to: [464139410](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464139410) [](ancestors = 464139410)